Made Node.print_tree() print prettily

sdfgeoff opened the pull request:
I found the output of print_tree() to be slightly obstuse to read for more than a few nodes. So I tweaked print_tree() to give a nicer output. This is extremely useful where objects are loaded dynamically and you need to see the whole project structure.

Example tree:
```
 '-root
    |-game
    |  |-LevelManager
    |  '-CharacterManager
    |     '-Player
    |        |-PlayerPhysics
    |        |  |-HeadCollision
    |        |  |-TorsoCollision
    |        |  '-PlayerGeometry
    |        |     |-Cylinder
    |        |     |-Cylinder_001
    |        |     |-Cylinder_002
    |        |     |-AnimationPlayer
    |        |     '-CameraTilt
    |        |        |-SpotLight
    |        |        '-Camera
    |        '-PreviewLight
    '-HUD
       |-Menu
       |  |-Label
       |  '-Camera2D
       '-SplashScreen
          '-Camera2D
```

I did want to use the box drawing characters, but for some reason they weren't printing.

Reasons why you may not want to include this pull request:
 - You can no longer copy/paste from print_tree() to the get_node() function.

Sslaxx commented:
Maybe having print_tree with a kind of "pretty printer" option to support both the default that allows copy/paste and this version?

LikeLakers2 commented:
Why not name this `Node.pretty_print_tree()`?

sdfgeoff commented:
That is one way to solve the conflict, but I would prefer fewer options then pretty printing - otherwise you get an expansion of features and no-one who understands them. If you've ever used ROS or OpenCV you'll see why I want to avoid feature explosion.

For example: who would bother to type "true" or "false" to separate between pretty and not pretty printing when this is a debug function used very seldom. The extra functionality of printing prettily is worth less (to me) than the additional mental load of having an option.

Noshyaar commented:
>when this is a debug function used very seldom

How about `#ifdef DEBUG_ENABLED` it so it works only in debug build?

LikeLakers2 commented:
@sdfgeoff I still disagree with entirely replacing how this outputs, even in debug builds. I would rather it be an option. My concern is as you said, that this new output can't just be pasted into get_node(), so why not make it an option so that people can choose between output that *can* be pasted into get_node(), or output that looks prettier?

vnen commented:
In any case, this breaks compatibility because the expected output changes completely, so it'd probably be better to add as new function.

sdfgeoff commented:
However, this output is what I would expect from a function called "print_**tree**" The existing behavior is what I would expect from "print_children(recursive=true)" if such a function exists.
Another alternative is to make this version print the full path as well, but I think that coupled with the indents, this would make things too long.

So, a quick poll (I'll only have time to work on it tomorrow anyway):

![Hooray Hat](https://assets-cdn.github.com/images/icons/emoji/unicode/1f389.png) for making it `pretty_print_tree()`

![Heart](https://assets-cdn.github.com/images/icons/emoji/unicode/2764.png) for making it `print_tree(pretty=false)`

![Laugh](https://assets-cdn.github.com/images/icons/emoji/unicode/1f604.png) for renaming existing `print_tree()` into `print_children(recursive=True)` (unlikely to be accepted as it would be an API change, but I'll put it here anyway)

sdfgeoff committed:
Made print_tree_pretty() function which displays scene tree graphically

sdfgeoff commented:
Well, the response was rather clear. I've updated it so it's `print_tree_pretty()`. I've gone for that word order so that it sorts in alphabetical order without issues. 

Documentation is also added. I added small examples to the docs to make it clearer what the output of two nearly identical functions is without the user having to run them.

silverkorn commented:
I'm a bit confused on your pick. You suggested all the way `pretty_print_tree()` but implemented `print_tree_pretty()` for better function sorting?

Afaik, the usual way to say it is _pretty_print_ rather than _print_pretty_.

Just to make sure about your decision :grin:

sdfgeoff commented:
I did implement it as pretty_tree_print first, and then when generating the docs, noticed that there were several other functions that appeared in-between "print_tree" and "pretty_print_tree". I decided that I would rather that they were grouped when sorted by name so that a person looking at the documentation can make a direct comparison between them.

akien-mga commented on doc/classes/Node.xml:
```diff
@@ -438,7 +438,30 @@
 			<description>
-				Prints the scene to stdout. Used mainly for debugging purposes.
+				Prints the tree to stdout. Used mainly for debugging purposes. This version displays the path relative to the current node, and is good for copy/pasting into the get_node() function.
+				[code]
```
You should use `[codeblock]` for code blocks, and indent the block to match the rest of the `<description>`. If you need further indentation in the code block itself (like for GDScript functions, use 4 spaces per indent).

akien-mga commented on doc/classes/Node.xml:
```diff
@@ -438,7 +438,30 @@
+			<return type="void">
+			</return>
+			<description>
+				Similar to print_tree, this prints the tree to stdout. This version displays a more graphical representation similar to what is displayed in the scene inspector. It is useful for inspecting larger trees.
```
Use `[method print_tree]` to make an hyperlink to the method you're referring to.

akien-mga commented on scene/main/node.cpp:
```diff
@@ -1868,18 +1868,31 @@ bool Node::has_persistent_groups() const {
-	print_line(String(p_node->get_path_to(this)));
-	for (int i = 0; i < data.children.size(); i++)
-		data.children[i]->_print_tree(p_node);
+	String new_prefix = last ? " '-" : " |-"; // I would use utf-8 box drawing characters....
```
Why not use UTF-8 box drawing characters? :)

sdfgeoff replied:
They weren't displaying for me in the Godot window. I'm not sure why.

Noshyaar replied:
Try `String::utf8(box_drawing_char)`

sdfgeoff replied:
Thanks @Noshyaar, that was what I needed to know.

akien-mga commented on scene/main/node.cpp:
```diff
@@ -1868,18 +1868,31 @@ bool Node::has_persistent_groups() const {
 	return false;
 }
-void Node::_print_tree(const Node *p_node) {
+void Node::_print_tree_pretty(const String prefix, const bool last) {
```
Are those arguments really useful if they're not exposed? Best keep it simple if they're not meant to be used.

sdfgeoff replied:
It's a recursive function, so while they're not exposed, they are needed: eg for indentation.

menip replied:
In that case, shouldn't you have a private helper function that does all the recursion?

sdfgeoff replied:
Hence the difference between print_tree_pretty and _print_tree_pretty. Or am I missing something?

reduz commented:
If conflicts are solved, this should be ready to merge

sdfgeoff commented:
I'll resolve this tomorrow.

hpvb commented:
Cherry picked into 3.0.3
